.github/workflows/ci.yaml

```diff
@@ -3,49 +3,73 @@ on: [push, pull_request]
   cache-invalidation-key: 0
+  nixpkgs-url: https://github.com/NixOS/nixpkgs/archive/b2737d4980a17cc2b7d600d7d0b32fd7333aca88.tar.gz
+
 jobs:
   cabal-test:
     name: cabal test
-    env:
-      NIXSHELL: nix-shell --pure
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v2.3.4
-    - uses: cachix/install-nix-action@v12
+    - uses: actions/checkout@v2
+    - uses: cachix/install-nix-action@v15
       with:
-        nix_path: nixpkgs=channel:nixos-unstable
-    - name: Build Nix dependencies
-      run: $NIXSHELL --run "echo '=== Installed ==='"
+        nix_path: "${{ env.nixpkgs-url }}"
     - name: Cache Cabal dependencies
       uses: actions/cache@v2
       with:
         path: |
           ~/.cabal/packages
           ~/.cabal/store
           dist-newstyle
-        key: cache-${{ runner.os }}-${{ hashFiles('nixpkgs.nix') }}-v${{ env.cache-invalidation-key }}-${{ hashFiles('linear-base.cabal') }}-${{ github.sha }}
-        restore-keys: cache-${{ runner.os }}-${{ hashFiles('nixpkgs.nix') }}-v${{ env.cache-invalidation-key }}-${{ hashFiles('linear-base.cabal') }}-
+        key: cabal-deps-${{ runner.os }}-${{ hashFiles('nix/sources.json') }}-v${{ env.cache-invalidation-key }}-${{ hashFiles('linear-base.cabal') }}-${{ github.sha }}
+        restore-keys: cabal-deps-${{ runner.os }}-${{ hashFiles('nix/sources.json') }}-v${{ env.cache-invalidation-key }}-${{ hashFiles('linear-base.cabal') }}-
+    - name: Build Nix dependencies
+      run: nix-shell --pure --run "echo '=== Nix dependencies installed ==='"
     - name: Update Cabal's database
-      run: $NIXSHELL --run "cabal update"
+      run: nix-shell --pure --run "cabal update"
     - name: Build Cabal's dependencies
-      run:  $NIXSHELL --run "cabal build --enable-tests --enable-benchmarks --dependencies-only"
+      run: nix-shell --pure --run "cabal build --enable-tests --enable-benchmarks --dependencies-only"
     - name: Build
-      run: $NIXSHELL --run "cabal build"
+      run: nix-shell --pure --run "cabal build"
     - name: Test
-      run: $NIXSHELL --run "cabal test"
+      run: nix-shell --pure --run "cabal test"
     - name: Haddock
-      run: $NIXSHELL --run "cabal haddock"
+      run: nix-shell --pure --run "cabal haddock"
+
+  ormolu:
+    name: check formatting with ormolu
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - uses: cachix/install-nix-action@v15
+      with:
+        nix_path: "${{ env.nixpkgs-url }}"
+    - name: Cache Stack dependencies
+      uses: actions/cache@v2
+      with:
+        path: ~/.stack
+        key: stack-deps-${{ runner.os }}-${{ hashFiles('nix/sources.json') }}-v${{ env.cache-invalidation-key }}-${{ hashFiles('stack.yaml.lock', 'linear-base.cabal') }}-${{ github.sha }}
+        restore-keys: stack-deps-${{ runner.os }}-${{ hashFiles('nix/sources.json') }}-v${{ env.cache-invalidation-key }}-${{ hashFiles('stack.yaml.lock', 'linear-base.cabal') }}-
```

Comment:
I think that if both the caches are called the same they will be shared, which probably won't have the intended effect. It's probably best to give this one a different name.

Reply:
I wanted to share the cache in that case. I didn't see reasons not to share it, but maybe there are?

Reply:
Well, in fact, you are right, it's better to separate the keys (otherwise, because the jobs are executed in parallel, and because ormolu is the slowest at the moment, the "stack build" job write its cache first, and the ormolu job then cannot upload its own cache)